File: api/src/test/java/org/pmiops/workbench/test/FakeClock.java
package org.pmiops.workbench.test;

import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;

/**
 * Mutable clock implementation for testing.
 */
public class FakeClock extends Clock {

  private Instant instant;
  private ZoneId zoneId;

  public FakeClock(Instant instant, ZoneId zoneId) {
    this.instant = instant;
    this.zoneId = zoneId;
  }

  public FakeClock(Instant instant) {
    this(instant, ZoneId.systemDefault());
  }

  public FakeClock() {
    this(Instant.now());
  }

  @Override
  public long millis() {
    return instant.toEpochMilli();
  }

  @Override
  public ZoneId getZone() {
    return zoneId;
  }

  @Override
  public Instant instant() {
    return instant;
  }

  @Override
  public Clock withZone(ZoneId zone) {
    return new FakeClock(instant, zone);
  }

  public void setInstant(Instant instant) {
    this.instant = instant;
  }

  public void increment(long millis) {
    setInstant(Instant.ofEpochMilli(this.instant.toEpochMilli() + millis));
  }
}

File: api/src/test/java/org/pmiops/workbench/test/FakeClock.kt
package org.pmiops.workbench.test

import java.time.Clock
import java.time.Instant
import java.time.ZoneId

/** Mutable clock implementation for testing.  */
class FakeClock @JvmOverloads constructor(private var instant: Instant? = Instant.now(), private val zoneId: ZoneId = ZoneId.systemDefault()) : Clock() {

    override fun millis(): Long {
        return instant!!.toEpochMilli()
    }

    override fun getZone(): ZoneId {
        return zoneId
    }

    override fun instant(): Instant? {
        return instant
    }

    override fun withZone(zone: ZoneId): Clock {
        return FakeClock(instant, zone)
    }

    fun setInstant(instant: Instant) {
        this.instant = instant
    }

    fun increment(millis: Long) {
        setInstant(Instant.ofEpochMilli(this.instant!!.toEpochMilli() + millis))
    }
}
